[function](date function)add new date function 'last_day'

lsy3993 opened the pull request:
# Proposed changes

Issue Number: close #13139

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [x] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

lsy3993 committed:
add nee date function 'last_day'

lsy3993 committed:
add new date function 'lsy_day'

zxealous committed:
[test](cache)Add remote cache ut (#13377)

cambyzju committed:
[fix](array-type) check value valid while insert data into array column (#13365)

We should prevent insert while value overflow.

1. create table:
`CREATE TABLE test_array_load_test_array_int_insert_db.test_array_load_test_array_int_insert_tb ( k1 int NULL, k2 array<int> NULL ) DUPLICATE KEY(k1) DISTRIBUTED BY HASH(k1) BUCKETS 5`

2. try insert data less than INT_MIN.
`insert into test_array_load_test_array_int_insert_tb values (1005, [-2147483649])`

Before this pr, the insert will success, but the value it not correct.
Co-authored-by: cambyzju <zhuxiaoli01@baidu.com>

xy720 committed:
[fix](array-type) Fix cast null to array make be core (#13324)

Doris do not support explicitly cast NULL_TYPE to ANY type .

```
mysql> select cast(NULL as int);
ERROR 1105 (HY000): errCode = 2, detailMessage = Invalid type cast of NULL from NULL_TYPE to INT
```

So we should also forbid user from casting NULL_TYPE to ARRAY type.

This commit will produce the following effect:

```
mysql> select cast(NULL as array<int>);
ERROR 1105 (HY000): errCode = 2, detailMessage = Invalid type cast of NULL from NULL_TYPE to ARRAY<INT(11)>
```

stalary commented on be/src/exprs/timestamp_functions.cpp:
```diff
@@ -707,6 +707,38 @@ DateTimeVal TimestampFunctions::from_days(FunctionContext* ctx, const IntVal& da
+
+    DateTimeValue ts_value = DateTimeValue::from_datetime_val(ts_val);
+
+    bool is_leap_year = (ts_value.year() % 400 == 0) ||
```
You can use doris::is_leap(year)

stalary replied:
Same as get_last_month_day?

lsy3993 replied:
done

stalary commented on be/src/vec/functions/function_timestamp.cpp:
```diff
@@ -569,6 +569,164 @@ class FunctionUnixTimestamp : public IFunction {
+    }
+
+    static int get_last_month_day(int year, int month) {
+        bool is_leap_year = (year % 400 == 0) || (year % 4 == 0 && year % 100 != 0);
```
Use doris::is_leap

lsy3993 replied:
done

xinyiZzz committed:
[fix](memory) Fix USE_JEMALLOC=true UBSAN compilation error #13398

liaoxin01 committed:
[fix](regression-test) fix that multiple cases conflict with the same table name (#13395)

hello-stephen committed:
[thirdparty](zstd)update dist info and thirdparty change log (#13392)

lsy3993 committed:
code format

HappenLee committed:
[Opt](vec) opt runtime filter for TPCH Q22 (#13339)

lsy3993 committed:
bug fix

Yukang-Lian committed:
[Feature](Retention) support retention function (#13056)

carlvinhust2012 committed:
[optimization](array-type) update the docs for import data to array column (#13345)

1. this pr is used to update the json load docs for  import data to array column.
when we use json to import data to array column,  the Rapidjson will cause precision problems. 
so we update the json-load docs to specify how to avoid these problems.

Issue Number: #7570
Co-authored-by: hucheng01 <hucheng01@baidu.com>

smallhibiscus committed:
[typo](docs) Fixed thrift_client_timeout_ms's incorrect description of en docs. (#13391)

Co-authored-by: smallhibiscus <8449081280@qq.com>

dataroaring committed:
[improvement](config) let default value of alter and load timeout suitable for most cases (#13370)

It is frustrated that a long running job fails due to small timeout. Actually, users
do not expect a timeout for a log running job.

xinyiZzz committed:
[enhancement](memtracker) Fix bthread local consume mem tracker (#13368)

Previously, bthread_getspecific was called every time bthread local was used. In the test at #10823, it was found that frequent calls to bthread_getspecific had performance problems.

So a cache is implemented on pthread local based on the btls key, but the btls key cannot correctly sense bthread switching.

So, based on bthread_self to get the bthread id to implement the cache.

luozenglin committed:
[feature](agg) support `group_bitmap_xor` agg function. (#13287)

support `group_bitmap_xor` agg function

xy720 committed:
[Enhancement](array-type) Tuple is null predicate support array type (#13307)

Issue Number: #12689

Gabriel39 commented on regression-test/suites/query_p0/sql_functions/datetime_functions/test_date_function.groovy:
```diff
@@ -424,4 +424,30 @@ suite("test_date_function") {
     qt_sql """ select seconds_sub(test_time2,1) result from ${tableName}; """
+
+    // test last_day
+    sql """ SET enable_vectorized_engine = TRUE; """
```
also add a case for row-based engine

lsy3993 replied:
done

Gabriel39 commented on be/src/vec/functions/function_timestamp.cpp:
```diff
@@ -569,6 +569,164 @@ class FunctionUnixTimestamp : public IFunction {
+
+    String get_name() const override { return name; }
+
+    bool use_default_implementation_for_nulls() const override { return true; }
```
`use_default_implementation_for_constants` also need to override and set to true

lsy3993 replied:
done

zhangstar333 commented on be/src/vec/functions/function_timestamp.cpp:
```diff
@@ -569,6 +569,164 @@ class FunctionUnixTimestamp : public IFunction {
+        } else {
+            // neither DateTime nor DateTimeV2/DateV2, return null
+            if constexpr (std::is_same_v<DateType, DataTypeDateTime>) {
+                res_column = ColumnInt64::create(input_rows_count);
```
this check seems to have been done at the above

lsy3993 replied:
you are right, I delete the 'else'

morningman committed:
[feature-wip](new-scan) Support stream load with csv in new scan framework (#13354)

1. Refactor the file reader creation in FileFactory, for simplicity.
    Previously, FileFactory had too many `create_file_reader` interfaces.
    Now unified into two categories: the interface used by the previous BrokerScanNode,
    and the interface used by the new FileScanNode.
    And separate the creation methods of readers that read `StreamLoadPipe` and other readers that read files.

2. Modify the StreamLoadPlanner on FE side to support using ExternalFileScanNode

3. Now for generic reader, the file reader will be created inside the reader, not passed from the outside.

4. Add some test cases for csv stream load, the behavior is same as the old broker scanner.

carlvinhust2012 committed:
[optimization](array-type) array_min/array_max function support the date/datetime type (#13407)

This pr is used to expand the supported data type for array_min/array_max function.
Before the change , the array_min/array_max function can't support the date/datetime type.
After the change, array_min/array_max function can support the date/datetime type.
Co-authored-by: hucheng01 <hucheng01@baidu.com>

HappenLee committed:
[Bug](funciton) fix repeat coredump when step is to long (#13408)

englefly committed:
[feature](nereids) let minValue and maxValue in stats support for Date, CHAR and VARCHAR type (#13311)

1. enable varchar/char type set min/max value.
    take first 8 chars as long, and convert to double.
2. fix bug when set min/max value for date and datav2

Gabriel39 committed:
[Improvement](join) optimize join probing phase (#13357)

Gabriel39 committed:
[Regression](javaudf) add regression test for javaudf (#13266)

adonis0147 committed:
[enhancement](macOS M1) Support building from source on macOS (M1) (#13195)

# Proposed changes

This PR fixed lots of issues when building from source on macOS with Apple M1 chip.

## ATTENTION

The job for supporting macOS with Apple M1 chip is too big and there are lots of unresolved issues during runtime:
1. Some errors with memory tracker occur when BE (RELEASE) starts.
2. Some UT cases fail.
...

Temporarily, the following changes are made on macOS to start BE successfully.
1. Disable memory tracker.
2. Use tcmalloc instead of jemalloc.

This PR kicks off the job. Guys who are interested in this job can continue to fix these runtime issues.

## Use case

```shell
./build.sh -j 8 --be --clean

cd output/be/bin
ulimit -n 60000
./start_be.sh --daemon
```

## Something else

It takes around _**10+**_ minutes to build BE (with prebuilt third-parties) on macOS with M1 chip. We will improve the  development experience on macOS greatly when we finish the adaptation job.

AshinGau committed:
[feature-wip](multi-catalog) use apache orc reader to read orc file (#13404)

Use apache orc to read orc file, and convert ColumnVectorBatch to doris block.

lsy3993 committed:
code optimize

lsy3993 commented on be/src/vec/functions/function_timestamp.cpp:
```diff
@@ -569,6 +569,164 @@ class FunctionUnixTimestamp : public IFunction {
+    // input DateTime and Date, return Date
+    // input DateTimeV2 and DateV2, return DateV2
+    DataTypePtr get_return_type_impl(const ColumnsWithTypeAndName& arguments) const override {
+        if constexpr (std::is_same_v<DateType, DataTypeDateTime>) {
```
done

yixiutt committed:
[improvement](compaction) delete num based compaction policy (#13409)

Co-authored-by: yixiutt <yixiu@selectdb.com>

weizhengte committed:
[feature-wip](statistics) collect statistics by sampling sql-tasks (#13399)

1. Collect statistics by sampling sql-tasks.
2. Consolidate statistics SQL statements and remove redundant statements.

lsy3993 committed:
Merge remote-tracking branch 'upstream/master'
